meeting-notes/2020/Q2/2020-05-21--design-discussion-subdomains-dns-label-limit.md

```diff
@@ -0,0 +1,93 @@
+# Design Discussion: Subdomain support for CIDs longer than 63 characters
+
+## Context
+
+Full explanation: https://github.com/ipfs/go-ipfs/issues/7318
+
+**tl;dr**  
+
+[RFC 1034](https://tools.ietf.org/html/rfc1034#page-7) states that "each node has a label, 
+which is zero to 63 octets in length", this means CID in subdomain can't be longer than 63 characters.
+
+If we split at 63rd character that will produce TLS error because it is impossible 
+to have TLS cert for more than one wildcard level (`*.*.ipns.dweb.link`), as clarified in [RFC6125](https://tools.ietf.org/html/rfc6125#page-27)
+and implemented by browser vendors.
+
+DNS length limits were known, but we did not include inlined ed25519 in evaluation at a time, and this recent development changes the timeline, so we need to have a fixes for longer CIDs and TLS for IPNS websites.
+
+The Goal of this meeting is to find the least invasive path forward.
+
+
+## Participants
+
+- @Stebalien
+- @lidel
+- @ribasushi
+- @aschmahmann
+
+
+
+## Options vs Tradeoffs
+
+### Generic fix
+
+Split after 63rd char: "Do it anyway" ?
+add support, if not for ed25519, it will be useful for longer CIDs (they won't have TLS anyway) 
+
+
+### IPNS/ed25519-specific fixes
+
+Anything we can do to give IPNS websites created with ed25519 TLS without warnings?
+
+| Fix    | TLS for IPNS | TLS for long CID  | touches libp2p | different text representation in subdomain than base32 |
+| --- | --- | --- | --- | --- |
+| custom multicodec for ed25519 keys | YES | NO | YES | NO |
+| Allow PeerIDs as multihash in base32 (only in `*.ipns` subdomains) | YES | NO | NO | YES |
+| PeerIDs as CIDv1 in base36 | YES | NO | YES | YES |
+| "sha256 split in two" idea [missing notes, sorry!] | YES | ? | NO | YES |
+
+## Notes
+
+- all agree that touching key encoding in libp2p too expensive atm
```

Comment:
Will this not be more expensive later?

Reply:
Maybe better way to ask: If we change this later, will it result in backwards incompatible things later, which will require wholesale migration? Or is the nature of this change such that we can do it anytime and doesn't require changes up and down the stack?

Reply:
The phrasing is a little ambiguous. From my PoV the consensus seems to be "we can't touch libp2p at this stage already - we are stuck with the key format forever"

Reply:
I am sorry, I did not do good job with writing this down.

The way I understand  this, it is "expensive", because adding any complexity to multicodec/cid parsers need to be paid in more works or slower adoption / flexibility. 

Here is improved version which provides some background why:

```suggestion
- changing the way we represent keys in libp2p is expensive and may introduce problems in the future
  - right now all key types represented as CIDs get `libp2p-key` codec
    - this makes it easy to add support for different key types in the future without the need for updating CID/multicodec-related libraries
  - if we introduce a custom multicodec for ed25519 this creates bad precedent of moving key type to multicodec layer: leaking abstractions and making it harder to support across implementations 
```